### ISSUE TITLE
Add boundary checking.

### ISSUE THREAD

**Issue by mrakgr** [DiffSharp/DiffSharp#17]
Literally every professionally done numerical library has this in some form. It caused me some hilarious mistakes when I first discovered DiffSharp and not having them for 'performance' is not a valid reason in my mind. This is one area where one should not skimp out on.

DiffSharp should definitely have had this from day one.

Edit: It also bears mentioning that in the Cuda version, as all the kernel launches are asynchronous, the CPU's job is literally to act as an accountant while it waits for the GPU to finish. It is no issue at all from the performance standpoint.

**Comment by gbaydin** [comment]
Thank you for this comment. As I said before, we will probably add boundary checking to some functions.

In the meantime, if you are experiencing any problem because of this, I would be very happy if you could share a concrete piece of your code where the lack of boundary checking causes incorrect behavior. If that happens, we should treat it as a bug.

**Comment by mrakgr** [comment]
In my [first issue](https://github.com/DiffSharp/DiffSharp/issues/8) here, I posted some code where I incorrectly sized the hidden to hidden layer.

    let createRandomLayer hidden_size input_size act =
        {
        W = DM.init hidden_size input_size (fun _ _ -> (rng.NextDouble()-0.5) / sqrt(float hidden_size) |> float32)
        U = DM.init hidden_size input_size (fun _ _ -> (rng.NextDouble()-0.5) / sqrt(float hidden_size) |> float32)
        b = DV.init hidden_size (fun _ -> (rng.NextDouble()-0.5) / sqrt(float hidden_size) |> float32)
        a = act
        }

`U` should be `hidden_size hidden_size`. Never threw an exception. In fact I do not think any of the BLAS functions ever throw exceptions for such cases. This is also an issue with Hadamarad products. Actually let's go with addition first.

    let a = DM.create 3 3 2.0f
    let b = DM.create 6 6 2.0f

    let c = a+b

This results in:

    val c : DM = DM [[4.0f; 4.0f; 4.0f; 4.0f; 4.0f; 4.0f]
                     [4.0f; 4.0f; 4.0f; 2.0f; 2.0f; 2.0f]
                     [2.0f; 2.0f; 2.0f; 2.0f; 2.0f; 2.0f]
                     [2.0f; 2.0f; 2.0f; 2.0f; 2.0f; 2.0f]
                     [2.0f; 2.0f; 2.0f; 2.0f; 2.0f; 2.0f]
                     [2.0f; 2.0f; 2.0f; 2.0f; 2.0f; 2.0f]]

Clearly this can't be what the user intended. Let me try `let c = a .* b`

    val c : DM = DM [[4.0f; 4.0f; 4.0f]
                     [4.0f; 4.0f; 4.0f]
                     [4.0f; 4.0f; 4.0f]]

Now the DM magically resized itself. Let me try `a*b`

    val c : DM =
      DM
        [[24.0f; 24.0f; 24.0f; 24.0f; 24.0f; 24.0f]
         [12.0f; 12.0f; 12.0f; 12.0f; 12.0f; 12.0f]
         [1.06421016e+21f; 1.06421016e+21f; 1.06421016e+21f; 1.06421016e+21f;
          1.06421016e+21f; 1.06421016e+21f]]

Again, no exceptions thrown anywhere. Here is some more:

    let a = DM.create 1 1 2.0f
    let b = DM.create 100 100 2.0f

    let c = a*b

    val c : DM =
      DM
        [[1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f;
          1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f; 1.01945958e+21f]]

No exception was thrown.

> In the meantime, if you are experiencing any problem because of this, I would be very happy if you could share a concrete piece of your code where the lack of boundary checking causes incorrect behavior. If it causes incorrect results, we should treat it as a bug.

Unless you meant that I should share a correctly written piece of code where lack of boundary checking causes problems, the examples are quite easy to find.

**Comment by gbaydin** [comment]
Thank you for sharing example code. Here and in the issue you referenced, when you said boundary checking, I assumed you were talking about index checking when accessing elements. You were talking about many issues together, I think I have missed this one.

Yes, we did not have any warnings to let the user know when they attempt an operation where the dimensions don't make sense.

I see your point, this is important. I've added dimension checks now.
